### PR TITLE
NEW Can set workstation on product. Develop "service" management in BOM

### PR DESCRIPTION
**Pull request by atm-lena**
# NEW|New 

Add default workstation to a service and service management in BOM.

- Ref  - Service label
- Amount
- Units (by default filled with the one configured on the service)
- If workstation module activated, a default Workstation column pre-filled with that of the associated service
- Total cost (tooltip explaining the calculation according to the case below)
=> If WK activated: quantity (converted into hours) x thm of the workstation
=> If WK disabled: quantity (converted into hours) x service cost price

The total and general unit cost of the BOM must be updated with the sum of the table of products + services

Service table in BOM : 

image

Default workstation in service card : 

image

### CODE REVIEW

**Comment by atm-lena** [comment]
@eldy

File: htdocs/bom/ajax/ajax.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+$idproduct = GETPOST('idproduct', 'int');
+
+
+/*
+ * View
+ */
+
+top_httphead();
+
+if ($action == 'getDurationUnitByProduct') {
```
Comment (eldy) [inline htdocs/bom/ajax/ajax.php:69]: For each action of an ajax service, we must add a condition on a permission so here
`if ($action == 'getDurationUnitByProduct' && $user->hasRight('product', 'lire') {`
for example

File: htdocs/core/lib/functions.lib.php [hunk 2]
```diff
@@ -8532,10 +8532,11 @@ function dol_osencode($str)
  * 		@param	string	$fieldkey		Field to search the key into
  * 		@param	string	$fieldid		Field to get
  *      @param  int		$entityfilter	Filter by entity
+ * 	    @param	string		$filters	Filter on other fields
  *      @return int						<0 if KO, Id of code if OK
  *      @see $langs->getLabelFromKey
  */
-function dol_getIdFromCode($db, $key, $tablename, $fieldkey = 'code', $fieldid = 'id', $entityfilter = 0)
+function dol_getIdFromCode($db, $key, $tablename, $fieldkey = 'code', $fieldid = 'id', $entityfilter = 0, $filters = array())
```
Comment (eldy) [inline htdocs/core/lib/functions.lib.php:8539]: $filters = array or string ?
Later we do 
$sql .= $filters;

File: htdocs/core/class/cunits.class.php [hunk 3]
```diff
@@ -420,15 +420,16 @@ public function delete($user, $notrigger = 0)
...
+	 * @param string $unit_type weight,size,surface,volume,qty,time...
 	 * @return int            <0 if KO, Id of code if OK
 	 */
-	public function getUnitFromCode($code, $mode = 'code')
+	public function getUnitFromCode($code, $mode = 'code', $unit_type = '')
 	{
 
 		if ($mode == 'short_label') {
-			return dol_getIdFromCode($this->db, $code, 'c_units', 'short_label', 'rowid');
+			return dol_getIdFromCode($this->db, $code, 'c_units', 'short_label', 'rowid', 0, ' AND unit_type = "'.$unit_type.'"');
```
Comment (eldy) [inline htdocs/core/class/cunits.class.php:430]: Because string is used "as it is" in sql, you must have it escaped for sql so
`' AND unit_type = "'.$this->db->escape($unit_type).'"');`

File: htdocs/core/class/cunits.class.php [hunk 4]
```diff
@@ -420,15 +420,16 @@ public function delete($user, $notrigger = 0)
...
 		} elseif ($mode == 'code') {
-			return dol_getIdFromCode($this->db, $code, 'c_units', 'code', 'rowid');
+			return dol_getIdFromCode($this->db, $code, 'c_units', 'code', 'rowid', 0, ' AND unit_type = "'. $unit_type .'"');
```
Comment (eldy) [inline htdocs/core/class/cunits.class.php:432]: Because string is used "as it is" in sql, you must have it escaped for sql so
' AND unit_type = "'.$this->db->escape($unit_type).'"');

File: htdocs/bom/tpl/objectline_view.tpl.php [hunk 5]
```diff
@@ -60,7 +65,7 @@
 }
 
 // add html5 elements
-$domData  = ' data-element="'.$line->element.'"';
+$domData  = ' data-element="'.$line->element.'service"';
```
Comment (eldy) [inline htdocs/bom/tpl/objectline_view.tpl.php:68]: Can you confirm the add of the string 'service' after $line->element ?

**Comment by atm-lena** [comment]
@eldy  fixed
